---
 libmultipath/discovery.c |    6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

Index: multipath-tools-696133/libmultipath/discovery.c
===================================================================
--- multipath-tools-696133.orig/libmultipath/discovery.c
+++ multipath-tools-696133/libmultipath/discovery.c
@@ -716,7 +716,7 @@ path_offline (struct path * pp)
 	char buff[SCSI_STATE_SIZE];
 
 	if (path_state(pp, buff))
-		return 1;
+		return 0;
 
 	if (!strncmp(buff, "offline", 7)) {
 		pp->offline = 1;
@@ -864,6 +864,10 @@ get_state (struct path * pp, int daemon)
 			return PATH_PENDING;
 		checker_set_async(c);
 	}
+	if (!pp->sysdev) {
+		condlog(2, "%s: no sysfs information", pp->dev);
+		return PATH_DOWN;
+	}
 	if (!conf->checker_timeout)
 		sysfs_get_timeout(pp->sysdev, &(c->timeout));
 	state = checker_check(c);
